torch/CMakeLists.txt

```diff
@@ -265,73 +183,145 @@ if (NOT NO_API)
     ${TORCH_SRC_DIR}/csrc/api/src/optim/sgd.cpp
-  )
+    )
+
 endif()
 
 add_library(torch SHARED ${TORCH_SRCS})
 
-# https://gcc.gnu.org/onlinedocs/gcc-4.0.3/gcc/Warning-Options.html
-target_compile_options(torch
-  PRIVATE
-  -Wall
-  -Wextra
-  -pedantic
-  -Wcast-align
-  -Wcast-qual
-  -Wctor-dtor-privacy
-  -Wdisabled-optimization
-  -Winit-self
-  -Wmissing-include-dirs
-  -Woverloaded-virtual
-  -Wsign-promo
-  -Wstrict-overflow=5
-  -Wundef
-  -fdiagnostics-show-option
-  -Wno-unused-parameter
-  -Wno-missing-braces # This warning is buggy
-  -Wno-unknown-pragmas)
-
-if ($ENV{WERROR})
+# until they can be unified, keep these lists synced with setup.py
+if(MSVC)
+  target_compile_options(torch PRIVATE
+    /MD
+    /Z7
+    /EHa
+    /DNOMINMAX
+    /wd4267
+    /wd4251
+    /wd4522
+    /wd4522
+    /wd4838
+    /wd4305
+    /wd4244
+    /wd4190
+    /wd4101
+    /wd4996
+    /wd4275
+    /bigobj
+    )
+else()
+  target_compile_options(torch PRIVATE
+    -std=c++11
+    -Wall
+    -Wextra
+    -Wno-unused-parameter
+    -Wno-missing-field-initializers
+    -Wno-write-strings
+#    -Wno-zero-length-array
+    -Wno-unknown-pragmas
+    # This is required for Python 2 declarations that are deprecated
+    # in 3.
+    -Wno-deprecated-declarations
+    # Python 2.6 requires -fno-strict-aliasing see
+    # http://legacy.python.org/dev/peps/pep-3123/ We also depend on it
+    # in our code (even Python 3).
+    -fno-strict-aliasing
+    # Clang has an unfixed bug leading to spurious missing braces
+    # warnings, see https://bugs.llvm.org/show_bug.cgi?id=21629
+    -Wno-missing-braces
+    )
+endif()
+
+if (MSVC)
+elseif ($ENV{WERROR})
   target_compile_options(torch PRIVATE -Werror)
 endif()
 
+if (MSVC)
+  target_link_libraries(torch onnx onnx_library)
+  target_compile_definitions(torch PRIVATE _THP_CORE)
+endif()
+
 target_link_libraries(torch
-  ${TORCH_CUDA_LIBRARIES}
-  ${CAFFE2_LIBRARY}
-  ${PROTOBUF_LIBRARY}
-  protobuf-nanopb
-)
+  caffe2_library
+  protobuf-nanopb)
+
 if(USE_CUDA)
-  if("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU")
-    target_link_libraries(torch -Wl,--no-as-needed ${CAFFE2_GPU_LIBRARY} -Wl,--as-needed)
+  if(MSVC)
+    set(TORCH_CUDA_LIBRARIES
+      ${NVTOOLEXT_HOME}/lib/x64/nvToolsExt64_1.lib
+      ${CUDA_LIBRARIES})
+    target_include_directories(torch PRIVATE "${NVTOOLEXT_HOME}/include")
+  elseif(APPLE)
+    set(TORCH_CUDA_LIBRARIES
+      ${CUDA_TOOLKIT_ROOT_DIR}/lib/libcudart.dylib
+      ${CUDA_TOOLKIT_ROOT_DIR}/lib/libnvrtc.dylib
+      ${CUDA_TOOLKIT_ROOT_DIR}/lib/libnvToolsExt.dylib
+      ${CUDA_LIBRARIES})
+    set_target_properties(torch PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
+  else()
+    set(TORCH_CUDA_LIBRARIES
+      ${CUDA_CUDA_LIB}
+      ${CUDA_NVRTC_LIB}
+      ${CUDA_TOOLKIT_ROOT_DIR}/lib64/libnvToolsExt.so
+      ${CUDA_LIBRARIES})
+  endif()
+
+  if(MSVC OR APPLE)
+    target_link_libraries(torch caffe2_gpu_library ${TORCH_CUDA_LIBRARIES})
```

Comment:
Why is it `TORCH_CUDA_LIBRARIES` on MSVC/Apple but `CUDA_LIBRARIES` on Linux?

Reply:
I'm unclear on the exact underlying situation, but linking against the additional libraries breaks the build in some of our CI environments. I don't remember the error, would have to kick of another test run to reproduce it.

note that we do need the extra libraries when building executables (see test_api and test_jit below)

Reply:
it causes this failure https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-cuda8-cudnn6-py3-build/14172/console

which had something to do with libtorch.so not being able to find libcuda.so

Reply:
OK... I think this merits further investigation, but it doesn't have to be in this PR. I'd recommend commenting about it.

Reply:
++ I'll add a comment